File: solutions/prospective/algorithms/CLRS/lua/kadane.lua
--[[
Kadane's Algorithm (Maximum Subarray Sum)

Time Complexity: O(n)
Space Complexity: O(1)

Finds the contiguous subarray which has the largest sum.
--]]
-- === MEMO START ===
function max_subarray(nums)
    local max_so_far = nums[1]
    local current_max = nums[1]
    
    for i = 2, #nums do
        current_max = math.max(nums[i], current_max + nums[i])
        max_so_far = math.max(max_so_far, current_max)
    end
    
    return max_so_far
end

File: solutions/prospective/algorithms/CLRS/c/kadane.c
/*
Kadane's Algorithm (Maximum Subarray Sum)

Time Complexity: O(n)
Space Complexity: O(1)

Finds the contiguous subarray which has the largest sum.
*/
// === MEMO START ===
int max_subarray(int *nums, int n) {
    int max_so_far = nums[0];
    int current_max = nums[0];

    for (int i = 1; i < n; i++) {
        if (nums[i] > current_max + nums[i]) {
            current_max = nums[i];
        } else {
            current_max = current_max + nums[i];
        }
        
        if (current_max > max_so_far) {
            max_so_far = current_max;
        }
    }

    return max_so_far;
}

